### PR TITLE
[nexus] Add stubs for image APIs

### PR DESCRIPTION
**Pull request by smklein**
First part of #734

### CODE REVIEW

File: nexus/src/external_api/params.rs [hunk 1]
```diff
@@ -229,6 +229,19 @@ pub struct DiskIdentifier {
...
+ */
+
+/// Create-time parameters for an
+/// [`Image`](omicron_common::api::external::Image)
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+pub struct ImageCreate {
+    /// common identifying metadata
+    #[serde(flatten)]
+    pub identity: IdentityMetadataCreateParams,
+}
```
Comment (smklein) [inline nexus/src/external_api/params.rs:243]: We need more here to actually be able to create the image.

Do we want this to be an enum implying "Please provide one of a {file, snapshot UUID, URL}"? Do we want that to all be done through the same endpoint?

File: common/src/sql/dbinit.sql [hunk 2]
```diff
@@ -370,6 +370,26 @@ CREATE INDEX ON omicron.public.disk (
...
+    id UUID PRIMARY KEY,
+    name STRING(63) NOT NULL,
+    description STRING(512) NOT NULL,
+    time_created TIMESTAMPTZ NOT NULL,
+    time_modified TIMESTAMPTZ NOT NULL,
+    /* Indicates that the object has been deleted */
+    time_deleted TIMESTAMPTZ,
+
+    project_id UUID,
+    volume_id UUID NOT NULL,
```
Comment (smklein) [inline common/src/sql/dbinit.sql:384]: Once the image has been constructed, it'll refer to a particular volume.

For the case of e.g. POST-ing an image by URL, would we immediately download + store the image as a part of invoking the POST request?

- If yes: Cool, this structure should work. We'll create a saga that does that work.
- If no: This structure may need to be more complex - we'll need to store the URL too.

File: common/src/sql/dbinit.sql [hunk 3]
```diff
@@ -370,6 +370,26 @@ CREATE INDEX ON omicron.public.disk (
 ) WHERE
     time_deleted IS NULL AND attach_instance_id IS NOT NULL;
 
+CREATE TABLE omicron.public.image (
```
Comment (smklein) [inline common/src/sql/dbinit.sql:373]: RFD 4 mentions the following fields which are not yet stored here (nor a part of the API):
- version
- digest
- status

File: common/src/sql/dbinit.sql (hunk 2 above)
Comment (jmpesp) [reply common/src/sql/dbinit.sql:384]: This is currently an unsolved problem, though that answer will likely be the same as others: some block migration task will move blocks from the read only parent to the volume region set(s), then remove the read only parent later. We can't gate the instance boot or image creation on that download though.

File: nexus/src/external_api/http_entrypoints.rs [hunk 4]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
     apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
 }
 
+/*
+ * Images
+ */
+
+/// List global images.
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1127]:
```suggestion
/** List global images
 * 
 * Returns a list of all the global images. Global images are returned sorted by creation date, with the most recent images appearing first.
 */
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 5]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+            .await?
+            .into_iter()
+            .map(|d| d.into())
+            .collect();
+        Ok(HttpResponseOk(ScanByName::results_page(&query, images)?))
+    };
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// Create a global image.
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1157]:
```suggestion
/** Create a global image
 *
 * Create a new global image.
 */
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 6]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// Path parameters for Image requests
+#[derive(Deserialize, JsonSchema)]
+struct GlobalImagePathParam {
+    image_name: Name,
+}
+
+/// Get a global image.
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1184]:
```suggestion
/** Get a global image
 * 
 * Returns the details of a specific global image.
 */
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 7]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+    let image_name = &path.image_name;
+    let handler = async {
+        let opctx = OpContext::for_external_api(&rqctx).await?;
+        let image = nexus.image_fetch(&opctx, &image_name).await?;
+        Ok(HttpResponseOk(image.into()))
+    };
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// Delete a global image.
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1206]:
```suggestion
/** Delete a global image
 * 
 * Permanently delete a global image. This operation cannot be undone. Any instances using the global image will continue to run, however new instances can not be created with this image.
 */
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 8]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+    let image_name = &path.image_name;
+    let handler = async {
+        let opctx = OpContext::for_external_api(&rqctx).await?;
+        nexus.delete_image(&opctx, &image_name).await?;
+        Ok(HttpResponseDeleted())
+    };
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// List images in a project.
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1228]:
```suggestion
/** List images
 *
 * List images in a project. The images are returned sorted by creation date, with the most recent images appearing first.
 */
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 9]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+#[endpoint {
+    method = GET,
+    path = "/organizations/{organization_name}/projects/{project_name}/images",
+    tags = ["images"],
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1232]:
```suggestion
    tags = ["projects"],
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 10]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+            .await?
+            .into_iter()
+            .map(|d| d.into())
+            .collect();
+        Ok(HttpResponseOk(ScanByName::results_page(&query, images)?))
+    };
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// Create an instance in a project
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1264]:
```suggestion
/** Create an image
 *
 * Create a new image in a project.
 *./
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 11]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+#[endpoint {
+    method = POST,
+    path = "/organizations/{organization_name}/projects/{project_name}/images",
+    tags = ["images"]
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1268]:
```suggestion
    tags = ["projects"]
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 12]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+
+/// Path parameters for Image requests
+#[derive(Deserialize, JsonSchema)]
+struct ImagePathParam {
+    organization_name: Name,
+    project_name: Name,
+    image_name: Name,
+}
+
+/// Get an image in a project.
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1304]:
```suggestion
/** Get an image
 *
 * Get the details of a specific image in a project.
 */
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 13]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+#[endpoint {
+    method = GET,
+    path = "/organizations/{organization_name}/projects/{project_name}/images/{image_name}",
+    tags = ["images"],
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1308]:
```suggestion
    tags = ["projects"],
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 14]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+                &project_name,
+                &image_name,
+            )
+            .await?;
+        Ok(HttpResponseOk(image.into()))
+    };
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// Delete an image from a project.
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1335]:
```suggestion
/** Delete an image
 *
 * Permanently delete an image from a project. This operation cannot be undone. Any instances in the project using the image will continue to run, however new instances can not be created with this image.
 */
```

File: nexus/src/external_api/http_entrypoints.rs [hunk 15]
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
...
+#[endpoint {
+    method = DELETE,
+    path = "/organizations/{organization_name}/projects/{project_name}/images/{image_name}",
+    tags = ["images"],
```
Comment (jessfraz) [inline nexus/src/external_api/http_entrypoints.rs:1339]:
```suggestion
    tags = ["projects"],
```

**Comment by jessfraz** [comment]
sorry if the fmt is off in my suggestions. Wrote them by hand but will align well with #773 when merged

File: nexus/src/external_api/http_entrypoints.rs (hunk 9 above)
Comment (smklein) [reply nexus/src/external_api/http_entrypoints.rs:1232]: Is there documentation on what tag should be associated with what endpoint? I think I might not be understanding this fully. I thought that an operation acting on:

```
path = "/organizations/{organization_name}/projects"
```

Would be tagged with `projects`, but an operation acting on:

```
path = "/organizations/{organization_name}/projects/{project_name}/foobar"
```

Would be tagged with `foobar(s)`.

Are *both* supposed to be tagged with `projects`? If we keep nesting, is it just `projects` all the way down?

**Comment by jessfraz** [comment]
Yeah that’s fair I missed 770, is there a way to grep replace of I can do by hand, the content is mostly what matters

**Comment by smklein** [comment]
Hey @jessfraz , I incorporated your recommended comments. Regarding the tags, I added documentation for "images" to `tag-config.json` - but lemme know if you still think I'm using the wrong tags!

**Comment by jessfraz** [comment]
Awesome, tags and docs look great!

File: nexus/src/external_api/params.rs (hunk 1 above)
Comment (smklein) [reply nexus/src/external_api/params.rs]: Added an enum of "snapshot UUID or URL". We can continue to add extra params as necessary.

File: common/src/sql/dbinit.sql (hunk 2 above)
Comment (smklein) [reply common/src/sql/dbinit.sql]: Okay. I'll leave this question open as an implementation detail for follow-up PRs

**Pull request merged**